src/Intervals.php

```diff
@@ -0,0 +1,369 @@
+                            if ((string) $c2 === (string) $c || $c2 === self::anyDev()) {
+                                $otherGroupMatches++;
+                            }
+
+                            // != x && == x cancel each other, make sure none of these appears in the output
+                            if ($c->getOperator() === '==' && $c2->getOperator() === '!=' && $c->getVersion() === $c2->getVersion()) {
+                                $blacklist[] = (string) $c;
+                                $blacklist[] = (string) $c2;
+                            }
+                        }
+                    }
+
+                    // only keep == constraints which appear in all conjunctive sub-constraints
+                    if ($otherGroupMatches === \count($devGroups) - 1) {
+                        $dev[(string) $c] = $c;
+                    }
+                }
+            }
+            foreach ($blacklist as $c) {
+                unset($dev[$c]);
+            }
+        }
+
+        $dev = array_values($dev);
+
+        if (count($intervalGroups) === 1) {
+            return array('intervals' => $intervalGroups[0], 'devConstraints' => $dev);
+        }
+
+        $borders = array();
+        foreach ($intervalGroups as $group) {
+            foreach ($group as $interval) {
+                $borders[] = array('version' => $interval['start']->getVersion(), 'operator' => $interval['start']->getOperator(), 'side' => 'start');
+                $borders[] = array('version' => $interval['end']->getVersion(), 'operator' => $interval['end']->getOperator(), 'side' => 'end');
+            }
+        }
+
+        $opSortOrder = self::$opSortOrder;
+        usort($borders, function ($a, $b) use ($opSortOrder) {
+            $order = version_compare($a['version'], $b['version']);
+            if ($order === 0) {
+                return $opSortOrder[$a['operator']] - $opSortOrder[$b['operator']];
+            }
+
+            return $order;
+        });
+
+        $activeIntervals = 0;
+        $intervals = array();
+        $index = 0;
+        $activationThreshold = $constraint->isConjunctive() ? \count($intervalGroups) : 1;
+        $active = false;
+        foreach ($borders as $border) {
+            if ($border['side'] === 'start') {
+                $activeIntervals++;
+            } else {
+                $activeIntervals--;
+            }
+            if (!$active && $activeIntervals >= $activationThreshold) {
+                $intervals[$index] = array('start' => new Constraint($border['operator'], $border['version']));
+                $active = true;
+            }
+            if ($active && $activeIntervals < $activationThreshold) {
+                $active = false;
+
+                // filter out invalid intervals like > x - <= x, or >= x - < x
+                if (
+                    version_compare($intervals[$index]['start']->getVersion(), $border['version'], '=')
```

Comment:
Is it necessary to use version_compare even with normalized versions for equality?

Reply:
With normalized constraints no it is not needed I believe. We do have some tests using non-normalized constraints, and technically nothing preventing you from creating Constraint instances with non-normalized versions. I don't know if we want to just document this as "voiding warranty" and drop the version_compare for perf, or keep supporting these edge cases.

Reply:
BTW the failing tests that triggered me to add this were https://github.com/composer/semver/blob/master/tests/Constraint/ConstraintTest.php#L494 where b2 and beta2 do not compare equally without version_compare. Constraint::matches uses version_compare so technically it is supported to some extent, even though in Composer we never create non-normalized constraints AFAIK.